src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -112,6 +113,71 @@ public void holonomicDrive(double forward, double strafe, double rotation) {
         setStates(kinematics.toSwerveModuleStates(speeds));
     }
 
+    /**
+     * Move the swerve in the specified direction, rotation and velocity.
+     *
+     * @param forward  the velocity on the X-axis. [m/s]
+     * @param strafe   the velocity on the Y-axis. [m/s]
+     * @param rotation the rotational velocity counter-clockwise positive. [rad/s]
+     */
+    public void defaultHolonomicDrive(double forward, double strafe, double rotation) {
+        ChassisSpeeds speeds = fieldOriented ?
+                ChassisSpeeds.fromFieldRelativeSpeeds(forward, strafe, rotation, Robot.getAngle()) :
+                new ChassisSpeeds(forward, strafe, rotation);
+        setStates(kinematics.toSwerveModuleStates(speeds));
+    }
+
+    /**
+     * Move the swerve in the specified direction, rotation and velocity, but the velocities are relative to the error of the modules.
+     *
+     * @param forward  the velocity on the X-axis. [m/s]
+     * @param strafe   the velocity on the Y-axis. [m/s]
+     * @param rotation the rotational velocity counter-clockwise positive. [rad/s]
+     */
+    public void errorRelativeHolonomicDrive(double forward, double strafe, double rotation) {
+        ChassisSpeeds speeds = fieldOriented ?
+                ChassisSpeeds.fromFieldRelativeSpeeds(forward, strafe, rotation, Robot.getAngle()) :
+                new ChassisSpeeds(forward, strafe, rotation);
+        errorRelativeSetStates(kinematics.toSwerveModuleStates(speeds));
+    }
+
+    /**
+     * Set the states of the modules,  but the velocities are relative to the error of the modules.
+     *
+     * @param states the states of the modules.
+     */
+    private void errorRelativeSetStates(SwerveModuleState[] states) {
+        for (SwerveModule module : modules) {
+            states[module.getWheel()] = SwerveModuleState.optimize(states[module.getWheel()], module.getAngle());
+            double diff = Utils.deadband(states[module.getWheel()].angle.minus(module.getAngle()).getRadians(),
+                    Constants.SwerveDrive.ANGLE_COSINE_DEADBAND);
+            module.setAngle(states[module.getWheel()].angle);
+            module.setVelocity(states[module.getWheel()].speedMetersPerSecond * Math.cos(diff));
+        }
+    }
+
+    /**
+     * Check whether all modules have reached their desired angles.
+     *
+     * @param forward  the velocity on the X-axis. [m/s]
+     * @param strafe   the velocity on the Y-axis. [m/s]
+     * @param rotation the rotational velocity counter-clockwise positive. [rad/s]
+     */
+    public boolean haveModulesReachedAngles(double forward, double strafe, double rotation) {
```

Comment:
If you use this function internally, you can just pass the states instead of calculating the velocities twice